Initial Classes

XedinUnknown opened the pull request:
Added `Exception` and `InvalidArgumentException` standards-compliant classes.

- [x] @XedinUnknown to add missing docs.

XedinUnknown committed:
Downgraded min requirements to PHP 5.3 and added dependency

Now depending on `dhii/exception-interface`.

XedinUnknown committed:
Added `Exception` and test

XedinUnknown committed:
Added `InvalidArgumentException` and test

XedinUnknown commented:
@mecha, could you review this PR please?

mecha commented on src/InvalidArgumentException.php:
```diff
@@ -0,0 +1,76 @@
+     */
+    protected $argument;
+
+    public function __construct($message = '', $code = 0, RootException $previous = null, $argument = null)
```
Documentation required.

mecha commented on src/InvalidArgumentException.php:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @return $this
+     */
+    protected function _setArgument($argument)
```
Parameter doc missing.

mecha commented on test/functional/InvalidArgumentExceptionTest.php:
```diff
@@ -0,0 +1,95 @@
+        } catch (RootException $e) {
+            $this->assertInstanceOf(static::TEST_SUBJECT_CLASSNAME, $subject, 'A valid instance of the test subject could not be created');
+            $this->assertInstanceOf('Dhii\\Exception\\InvalidArgumentExceptionInterface', $subject, 'Subject is not a valid Dhii throwable');
+            $this->assertEquals($message, $e->getMessage(), 'Subject message is wrong');
```
I think an argument can be made that. given that this is a functional test, these assertions are not required as they have already been included in `ExceptionTest`.

XedinUnknown replied:
I thought about that. But on the other hand, this is a different concrete class, with a different constructor, which sets the internal properties, and therefore needs to be tested separately.

XedinUnknown commented:
- [ ] @XedinUnknown to add missing docs.

XedinUnknown committed:
Interface package stability increased

All deps updated

XedinUnknown committed:
Added test for `ArgumentAwareTrait`

XedinUnknown committed:
Added `ArgumentAwareTrait`

XedinUnknown committed:
`InvalidArgumentException` now uses trait for argument awareness

XedinUnknown committed:
Constructor of `InvalidArgumentException` is now standards-compliant

- Message can be `Stringable`
- All optional parameters default to `null`
- Documentation added

XedinUnknown committed:
`Exception` now standards compliant

Accepts `Stringable` along with string, and all constructor params default to `null`

XedinUnknown committed:
Tightened tests to test constructor defaults

XedinUnknown committed:
Standardized tests

XedinUnknown committed:
`InvalidArgumentException` now calls parameter-less constructor

XedinUnknown committed:
`Exception` now calls the new parameter-less constructor

XedinUnknown committed:
Standardized readme

XedinUnknown committed:
Added more verbosity to readme

mecha reviewed (approved):
LGTM
